<?xml version="1.0"?>
<launch>
  <!-- Lane Detection Node Launch File -->

  <!-- Arguments -->
  <arg name="camera_topic" default="/camera0/usb_cam_node/image_raw"/>
  <arg name="output_prefix" default="/camera0/lane_"/>
  <arg name="config_path" default="$(find LidarProjectionLane)/config/network_params/yolov8n_seg_lane.yaml"/>
  <arg name="enable_debug" default="true"/>
  <arg name="enable_performance_monitoring" default="true"/>
  <arg name="target_fps" default="30.0"/>

  <!-- Lane Detection Node -->
  <node name="lane_detection_node" pkg="LidarProjectionLane" type="lane_detection_node" output="screen">
    <!-- Topic Configuration -->
    <param name="input_topic" value="$(arg camera_topic)"/>
    <param name="output_topic_prefix" value="$(arg output_prefix)"/>
    <param name="frame_id" value="camera0_link"/>

    <!-- Model Configuration -->
    <param name="config_path" value="$(arg config_path)"/>

    <!-- Processing Configuration -->
    <param name="enable_debug_output" value="$(arg enable_debug)"/>
    <param name="enable_performance_monitoring" value="$(arg enable_performance_monitoring)"/>
    <param name="target_fps" value="$(arg target_fps)"/>

    <!-- YOLO Model Parameters -->
    <param name="yolo/confidence_threshold" value="0.5"/>
    <param name="yolo/nms_threshold" value="0.45"/>
    <param name="yolo/max_detections" value="100"/>
    <param name="yolo/input_width" value="416"/>
    <param name="yolo/input_height" value="416"/>
    <param name="yolo/tensorrt_precision" value="FP16"/>
    <param name="yolo/batch_size" value="1"/>
    <param name="yolo/workspace_size" value="1073741824"/>

    <!-- Lane Segmentation Parameters -->
    <param name="segmentation/min_area_threshold" value="100"/>
    <param name="segmentation/max_area_threshold" value="50000"/>
    <param name="segmentation/contour_epsilon" value="2.0"/>
    <param name="segmentation/morph_kernel_size" value="3"/>
    <param name="segmentation/morph_iterations" value="2"/>
    <param name="segmentation/hough_threshold" value="50"/>
    <param name="segmentation/min_line_length" value="30.0"/>
    <param name="segmentation/max_line_gap" value="10.0"/>
    <param name="segmentation/min_slope_threshold" value="0.3"/>
    <param name="segmentation/max_slope_threshold" value="3.0"/>

    <!-- Visualization Parameters -->
    <param name="visualization/left_lane_color_r" value="0"/>
    <param name="visualization/left_lane_color_g" value="255"/>
    <param name="visualization/left_lane_color_b" value="0"/>
    <param name="visualization/right_lane_color_r" value="0"/>
    <param name="visualization/right_lane_color_g" value="0"/>
    <param name="visualization/right_lane_color_b" value="255"/>
    <param name="visualization/center_lane_color_r" value="255"/>
    <param name="visualization/center_lane_color_g" value="255"/>
    <param name="visualization/center_lane_color_b" value="0"/>
  </node>

  <!-- Optional: Launch RViz for visualization -->
  <group if="$(arg enable_debug)">
    <node name="rviz_lane_detection" pkg="rviz" type="rviz"
          args="-d $(find LidarProjectionLane)/config/rviz/lane_detection.rviz"
          required="false"/>
  </group>

</launch>